lib/coach/middleware_validator.rb

```diff
@@ -21,14 +22,18 @@ def validated_provides!
 
     private
 
+    attr_reader :previous_middlewares
+
     def missing_requirements
-      @middleware.requirements - provided_by_chain
+      @missing_requirements ||= @middleware.requirements - provided_by_chain
     end
 
     def provided_by_chain
       @provided_by_chain ||=
         middleware_dependencies.reduce(@already_provided) do |provided, middleware|
-          provided + self.class.new(middleware, provided).validated_provides!
+          validator = self.class.new(middleware, previous_middlewares, provided)
+          previous_middlewares << middleware
```

Comment:
This side-effectiness makes me a bit sad. Could we just have:
```ruby
def previous_middlewares
  some_method_name.second
end

def provided_by_chain
  some_method_name.first
end

def some_method_name
  initial = [@already_provided, @previous_middlewares]
  middleware_dependencies.reduce(initial) do |(provided, previous), middleware|
    validator = self.class.new(middleware, previous, provided)
    [provided + validator.validated_provides!, previous + [middleware]]
  end
end
```

I think this would work?